---
title: Introduction to tSVE
author:
    -   name: KÃ©vin Rue-Albrecht
        email: k.rue-albrecht@imperial.ac.uk
        affiliation: Department of Medicine, Imperial College London,
            Hammersmith Campus, Du Cane Road, London, W12 0NN, UK
date: "`r doc_date()`"
package: "`r pkg_ver('tSVE')`"
abstract:
    The Shiny Variant Explorer (*tSVE*) provides functions to summarise and
    visualise gene-centric genetic variation data stored in VCF files
    pre-processed by the Ensembl Variant Effect Predictor (VEP).
    A Shiny web-application provides a convenient interface to demonstrate
    those functionalities in a programming-free environment.
vignette: >
    %\VignetteIndexEntry{Introduction to tSVE}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
output: 
    BiocStyle::pdf_document2
bibliography:
    Introduction.bib
---

# Introduction

The Shiny Variant Explorer (*tSVE*) offers functions to summarise and visualise
gene-centric genetic variation data stored in VCF files pre-processed by the
Ensembl Variant Effect Predictor (VEP) [@RN1].
A Shiny web-application provides a convenient interface to demonstrate those
functionalities in a programming-free environment.

# Installation

The Shiny Variant Explorer (`r Biocpkg("tSVE")`) can be installed using the
following code:

```{r biocLite, message=FALSE}
if (!require(tSVE)){
    source("http://bioconductor.org/biocLite.R")
    biocLite("tSVE")
}
```

\bioccomment{Use the `biocLite` function alone to update the package.}

Once installed, the package functions can be loaded as follows:

```{r library, message=FALSE}
library(tSVE)
```

# Recurrent settings: tSVEParam

Most functionalities in `r Biocpkg("tSVE")` require recurrent information
such as:

* genotypes (homozygote reference, heterozygote, homozygote alternate)
* INFO field containing the Ensembl VEP predictions
* suffixes of INFO fields where counts and frequencies of genotypes are stored
    + counts and frequencies may be calculated for levels of phenotypes,
        in which case the corresponding INFO fields are formed as
        `<phenotype>_<level>_<suffix>`
    + counts and frequencies across all samples are stored in INFO fields
        simply named `<suffix>`
* settings for parallel calculations

To reduce the burden of repetitive programming
and to 
facilitate analyses using consistent sets of parameters,
`r Biocpkg("tSVE")` implements the `tSVEParam` class.
`tSVEParam` objects offer a container for parameters recurrently used across
the package.
A `tSVEParam` object may be initialised as follows:

```{r tSVEParam}
tParam <- tSVEParam(
    genos = list(
        REF = c("0|0"),
        HET = c("0|1", "1|0"),
        ALT = c("1|1")),
    aaf = "AAF", # default
    maf = "MAF", # default
    vep = "CSQ", # default
    bp = BiocParallel::SerialParam()) # default
```

In this example:

* **Genotypes**
    + homozygote reference genotype is encoded `0|0`
        - counts of genotypes are stored in INFO fields suffixed with `REF`
    + heterozygote genotypes are encoded either `0|1` or `1|0`
        - counts of genotypes are stored in INFO fields suffixed with `HET`
    + homozygote alternate  genotype is encoded `1|1`
        - counts of genotypes are stored in INFO fields suffixed with `ALT`
* **Alternate allele frequency**
    + stored in INFO fields suffixed with `AAF`
* **Minor allele frequency**
    + stored in INFO fields suffixed with `MAF`
* **Ensembl VEP predictions**
    + imported from INFO field `CSQ`
* **Parallel calculations**
    + turned off. Serial calculations are performed.

Default values are provided for all slots except genotypes, as these may vary
more frequently from one data set to another.
\fixme{
In the future, however, a small set of pre-configured `tSVEParam` objects
may be included in the package for the most common situations
(e.g., `tSVEParamPhased`, `tSVEParamUnphased`, `tSVEParam012`).
}

# Data import

## Variants and phenotypes

Functionalities in `r Biocpkg("tSVE")` currently only use
`ExpandedVCF` objects of the `r Biocpkg("VariantAnnotation")` package,
which encode a single alternate allele per record.
`ExpandedVCF` objects can be produced:

* from VCF files
using the `r Biocpkg("VariantAnnotation")` `readVcf()` method
with argument `collapsed=FALSE`,
* from `r Biocpkg("VariantAnnotation")` `CollapsedVCF` objects
using the `r Biocpkg("VariantAnnotation")` `expand()` method.

To enable the key functionalities of the `r Biocpkg("tSVE")` package, the
`ExpandedVCF` object should include at least the following information:

* `fixed(x)`
    + `ALT`: alternate allele(s)
    
* `info(x)`
    + `<CSQ>`: where `<CSQ>` stands for a user-defined INFO field that
        contains Ensembl VEP predictions
    
* `geno(x)`
    + `GT`: genotypes

* `colData(x)`: phenotypes

## The preprocessVariants() method

The `r Biocpkg("tSVE")` method `preprocessVariants` offers an integrated route
to produce a minimal `ExpandedVCF` object that contains all the data
necessary for subsequent analyses.
To initialise a minimal `ExpandedVCF` object,
the `preprocessVariants` method imports:

* phenotypes
    + in any of the following formats:
        - `DataFrame` from the `S4Vectors` package
        - `data.frame`
        -  text file
    + in all cases, phenotypes must include:
        - `rownames`: sample identifiers
        - `colnames`: phenotype names
* variants within given genomic region(s)
    + `CHROM`, `POS`, `REF`, `ALT` fixed fields
    + Ensembl VEP predictions from a given INFO field for the above variants
        - Default: `CSQ`
    + genotypes (`GT`) for samples identified in the above phenotypes
        `rownames`
        - if no phenotypes are given, all samples are imported from the VCF
            file
    
The imported variants are expanded into bi-allelic records
(*i.e.,* one reference allele and one alternate allele per record),
and phenotypes are combined as the `colData` of the resulting `ExpandedVCF`
object.

## Example

Using the sample files included in the package, an `ExpandedVCF` object
can be obtained as follows:

```{r preprocessVariants, message=FALSE}
# Import phenotypes ----
phenoFile <- system.file(
    "extdata/integrated_samples.txt",
    package = "tSVE")

# Define VCF file ----
vcfFile <- system.file(
    "extdata/chr15.phase3_integrated.vcf.gz",
    package = "tSVE")

tabixVcf <- Rsamtools::TabixFile(file = vcfFile)

# Define genomic region(s) from which to import variants ----
library(GenomicRanges)
gr <- GenomicRanges::GRanges(
    seqnames = "15",
    IRanges::IRanges(
        start = 48413170,
        end = 48434757,
        names = "SLC24A5"))

# Import variants and phenotypes ----
vcf <- preprocessVariants(
    file = tabixVcf,
    regions = gr,
    param = tParam,
    phenos = phenoFile)
```

\bioccomment{
For `preprocessVariants`, `tSVEParam` objects are used
only for the `csq` value, that defines the INFO field
from which to extract Ensembl VEP predictions.
}

The result is an `ExpandedVCF` object including 481 variants in 2,504 samples:

```{r vcf, echo=FALSE}
vcf
```

# Summarising Ensembl VEP predictions

Although later sections discuss:

* the calculation of additional `info` fields (\fixme{Coming soon!}) 
* the filtering of variants based on `fixed`, `info`, and `geno` fields
    (\fixme{Coming soon!})

interesting figures and summary tables may be obtained as soon as the first
`ExpandedVCF` object is created.

Currently, the major `r Biocpkg("tSVE")` functionalities include:

* Counts of discrete Ensembl VEP predictions in levels of a phenotype
    + as a `data.frame` or a stacked \CRANpkg{ggplot2} `geom_bar`
    + counts faceted by another Ensembl VEP prediction
    + `geom_bar` represented as stacked percentage of counts
* Density distribution of continuous Ensembl VEP predictions in levels of
    a phenotype
    + density faceted by another Ensembl VEP prediction

## Counts of discrete Ensembl VEP predictions

### Barplot

The following chunk of code displays,
for each affected genomic feature,
the count of each Ensembl VEP prediction
`"Consequence"` in each "super population"
(*i.e.*, each level of the `"super_pop"` phenotype).

<!-- Cannot split chunk options over multiple lines :'-( -->
```{r tCBPplot,  fig.width=11, fig.height=6, out.extra='angle=90', out.width='6in', out.height='11in'}
tCBPplot <- tabulateCsqByPhenotype(
    vcf = vcf, phenoCol = "super_pop", csqCol = "Consequence",
    param = tParam,
    unique = TRUE, facet = "Feature", percentage = FALSE,
    plot = TRUE) +
    ggplot2::theme(
        legend.text = ggplot2::element_text(
            size = ggplot2::rel(0.5)),
        strip.text = ggplot2::element_text(
            size = ggplot2::rel(0.5)),
        axis.text.x = ggplot2::element_text(
            angle = 270,
            hjust = 0.5),
        axis.title.y = ggplot2::element_text(
            angle = 270)
    )

tCBPplot
```

### Table (data.frame)

The same counts can also be summarised as a `data.frame`:

```{r tCBPtableWide, eval=TRUE}
tCBPwide <- tabulateCsqByPhenotype(
        vcf = vcf, phenoCol = "super_pop", csqCol = "Consequence",
        param = tParam,
        unique = TRUE, facet = "Feature", plot = FALSE
    )
```

From which only the subset corresponding to *missense variants* is shown here:

```{r tCBPtableWideSubset, eval=FALSE}
subset(tCBPwide, Consequence == "missense_variant")
```

```{r tCBPtableWideSubsetShow, results="asis", echo=FALSE}
knitr::kable(
    x = subset(tCBPwide, Consequence == "missense_variant"),
    row.names = FALSE,
    caption = "Table of consequence by phenotype by feature in wide format")
```

## Distribution of continuous Ensembl VEP predictions

For Ensembl VEP prediction on a continuous scale, it may be desirable to
visualise the distribution of values as a density plot:

```{r dCBPplot}
densityCsqByPhenotype(
        vcf = vcf, phenoCol = "super_pop", csqCol = "CADD_PHRED",
        param = tParam,
        unique = TRUE, facet = "Feature", plot = TRUE, popFreq = FALSE)
```


# Session info

Here is the output of `sessionInfo()` on the system on which this document was
compiled:

```{r sessionInfo, echo=FALSE}
sessionInfo()
```

# References

[R]: http://r-project.org
[RStudio]: http://www.rstudio.com/
